src/value/builders.rs

```diff
@@ -0,0 +1,453 @@
+        let mut new_elements = Vec::with_capacity(elements.len());
+        new_elements.extend_from_slice(elements);
+        Self {
+            values: new_elements,
+        }
+    }
+
+    /// Adds the provided element to the end of the [`List`] being constructed.
+    pub fn push<E: Into<Element>>(mut self, element: E) -> Self {
+        self.values.push(element.into());
+        self
+    }
+
+    /// Removes the element at the specified position from the [`List`] being constructed.
+    /// If the index is out of bounds, this method will panic.
+    pub fn remove(mut self, index: usize) -> Self {
+        // This has O(n) behavior; the removals could be
+        // buffered until the build() if needed.
+        self.values.remove(index);
+        self
+    }
+
+    /// Builds a [`List`] with the previously specified elements.
+    pub fn build_list(self) -> List {
+        List::new(self.values)
+    }
+
+    /// Builds a list [`Element`] with the previously specified child elements.
+    pub fn build(self) -> Element {
+        self.build_list().into()
+    }
+}
+
+/// Constructs [SExp] values incrementally.
+///
+/// ```
+/// use ion_rs::ion;
+/// use ion_rs::value::owned::Element;
+/// let actual: Element = Element::sexp_builder()
+///     .push(1)
+///     .push(true)
+///     .push("foo")
+///     .build();
+/// let expected = ion!(r#"(1 true "foo")"#);
+/// assert_eq!(actual, expected);
+/// ```
+pub struct SExpBuilder {
+    values: Vec<Element>,
+}
+
+impl SExpBuilder {
+    /// Crate visible; users should call [`SExp::builder()`] or [`Element::sexp_builder()`] instead.
+    pub(crate) fn new() -> Self {
+        Self { values: Vec::new() }
+    }
+
+    /// Helper method for [`SExp::clone_builder()`].
+    pub(crate) fn with_initial_elements(elements: &[Element]) -> Self {
+        let mut new_elements = Vec::with_capacity(elements.len());
+        new_elements.extend_from_slice(elements);
+        Self {
+            values: new_elements,
+        }
+    }
+
+    /// Adds the provided element to the end of the [`SExp`] being constructed.
+    pub fn push<E: Into<Element>>(mut self, element: E) -> Self {
+        self.values.push(element.into());
+        self
+    }
+
+    /// Removes the element at the specified position from the [`SExp`] being constructed.
+    /// If the index is out of bounds, this method will panic.
+    pub fn remove(mut self, index: usize) -> Self {
+        // This has O(n) behavior; the removals could be
+        // buffered until the build() if needed.
+        self.values.remove(index);
+        self
+    }
+
+    /// Builds a [`SExp`] with the previously specified elements.
+    pub fn build_sexp(self) -> SExp {
+        SExp::new(self.values)
+    }
+
+    /// Builds an s-expression [`Element`] with the previously specified child elements.
+    pub fn build(self) -> Element {
+        self.build_sexp().into()
+    }
+}
+
+/// Constructs [Struct] values incrementally.
```

Comment:
> We should have an example showing how to use this to merge two structs (unless there's a different way that is better).

Sure, I'll add that.

> Is it worth adding a with_fields() or some other function to make it easier to add all the fields from one struct to this builder?

For the moment, this is achieved with `clone_builder()`:

```rust
let e1 = ion_struct! {"a": 1, "b": 2};
let e2 = e1.clone_builder().with_field("c", 3); // {a: 1, b: 2, c: 3}
```

We can/should add more sophisticated methods. (Currently there's no method to explicitly overwrite a given field; you'd have to remove the old one and then add the new one. Not ideal.) If it's ok, I'd like to do this in another PR since this one's already pretty big.

Reply:
>> Is it worth adding a with_fields() or some other function to make it easier to add all the fields from one struct to this builder?
>
>For the moment, this is achieved with clone_builder():

I should have been more clear. Should we have a better way to write something like this? (I don't actually know if this code will compile, but I think you can understand my meaning.)
```
let struct_a = ion_struct! { a: 1, b: 2 };
let struct_b = ion_struct! { c: 3, d: 4 };
let merged_struct: Element = {
    let builder = struct_a.clone_builder();
    for (fname, value) in struct_b.fields() {
        builder.with_field(fname, value);
    }
    builder.build()
}
println!("{merged_struct}"); // prints { a:1, b:2, c:3, d:4 }
```

Reply:
We should have a better way to, yes. Is it ok if I add it in a future PR? If so, I'll open an issue to track it.

Reply:
I went ahead and added a `StructBuilder::with_fields` method.
```rust
let s1 = ion_struct! {foo: 1};
let s2 = ion_struct! {bar: 2};
let merged = Struct::builder()
  .with_fields(s1.fields())
  .with_fields(s2.fields())
  .build();
```
I have one more trait impl I need to do for Struct (`IntoIterator`) and then it'll look like this:
```rust
let s1 = ion_struct! {foo: 1};
let s2 = ion_struct! {bar: 2};
let merged = Struct::builder()
  .with_fields(&s1)
  .with_fields(&s2)
  .build();
```

Reply:
> Is it ok if I add it in a future PR? If so, I'll open an issue to track it.

Yes, that's fine of course. Sorry that I didn't see this sooner. If you want to leave any out _any_ suggestions to avoid scope creep in this PR, that's fine by me.